mirrord-layer/src/sockets.rs

```diff
@@ -1,240 +1,368 @@
-        let sockets = self.connections.lock().unwrap();
-        sockets.get(&sockfd).map(|socket| socket.address)
+impl Borrow<RawFd> for Socket {
+    fn borrow(&self) -> &RawFd {
+        &self.fd
     }
+}
+
+#[derive(Default)]
+struct State {
+    sockets: HashSet<Socket>,
+}
 
-    pub fn get_data_socket_address(&self, sockfd: SockFd) -> Option<SocketAddr> {
-        let sockets = self.data.lock().unwrap();
-        sockets.get(&sockfd).map(|socket| socket.address)
+lazy_static! {
+    static ref STATE: Mutex<State> = Mutex::new(State::default());
+}
+
+#[inline]
+fn is_ignored_port(port: Port) -> bool {
+    port == 0 || (port > 50000 && port < 60000)
+}
+
+fn socket(domain: c_int, type_: c_int, protocol: c_int) -> RawFd {
+    debug!("socket called domain:{:?}, type:{:?}", domain, type_);
+    let fd = unsafe { libc::socket(domain, type_, protocol) };
+    if fd == -1 {
+        error!("socket failed");
+        return fd;
     }
+    // We don't handle non TCPv4 sockets
+    if !((domain == libc::AF_INET) || (domain == libc::AF_INET6) && (type_ & libc::SOCK_STREAM) > 0)
+    {
+        debug!("non TCP socket domain:{:?}, type:{:?}", domain, type_);
+        return fd;
+    }
+    let mut state = STATE.lock().unwrap();
+    state.sockets.insert(Socket {
+        fd,
+        domain,
+        type_,
+        protocol,
+        state: SocketState::default(),
+    });
+    fd
+}
+
+unsafe extern "C" fn socket_detour(domain: c_int, type_: c_int, protocol: c_int) -> c_int {
+    socket(domain, type_, protocol)
+}
 
-    pub fn read_single_connection(&self, sockfd: SockFd) -> ConnectionId {
-        let mut sockets = self.connections.lock().unwrap();
-
-        if let Some(mut socket) = sockets.remove(&sockfd) {
-            let mut buffer = [0; 1];
-            socket.read_socket.read_exact(&mut buffer).unwrap();
-            sockets.insert(sockfd, socket.address.port(), socket);
-            let mut queues = self.connection_queues.lock().unwrap();
-            let queue = queues.get_mut(&sockfd).unwrap();
-            queue.remove().unwrap()
-        } else {
-            error!("No socket found for fd: {}", sockfd);
-            0
+fn bind(sockfd: c_int, addr: *const sockaddr, addrlen: socklen_t) -> c_int {
+    debug!("bind called sockfd: {:?}", sockfd);
+    let mut socket = {
+        let mut state = STATE.lock().unwrap();
+        match state.sockets.take(&sockfd) {
+            Some(socket) if !matches!(socket.state, SocketState::Initialized) => {
+                error!("socket is in invalid state for bind {:?}", socket.state);
+                return libc::EINVAL;
+            }
+            Some(socket) => socket,
+            None => {
+                debug!("bind: no socket found for fd: {}", &sockfd);
+                return unsafe { libc::bind(sockfd, addr, addrlen) };
+            }
+        }
+    };
+
+    let raw_addr = unsafe { OsSocketAddr::from_raw_parts(addr as *const u8, addrlen as usize) };
```

Comment:
Why not move this to the beginning of the function? Then you could edit the socket as you find it with get_mut and not have to lock the State twice

Reply:
Because we want to remove it anway (if ignore we want to remove it,?